pkg/infrastructure/aws/clusterapi/iam.go

```diff
@@ -0,0 +1,242 @@
+					iamv1.PrincipalService: []string{
+						getPartitionService(ic.AWS.Region),
+					},
+				},
+				Action: iamv1.Actions{
+					"sts:AssumeRole",
+				},
+			},
+		},
+	}
+	assumePolicyBytes, err := json.Marshal(assumePolicy)
+	if err != nil {
+		return fmt.Errorf("failed to marshal assume policy: %w", err)
+	}
+
+	for _, role := range []string{master, worker} {
+		roleName, err := getOrCreateIAMRole(ctx, role, infraID, string(assumePolicyBytes), *ic, tags, svc)
+
+		// Put the policy inline.
+		policyName := aws.String(fmt.Sprintf("%s-%s-policy", infraID, role))
+		b, err := json.Marshal(policies[role])
+		if err != nil {
+			return fmt.Errorf("failed to marshal %s policy: %w", role, err)
+		}
+		if _, err := svc.PutRolePolicyWithContext(ctx, &iam.PutRolePolicyInput{
+			PolicyDocument: aws.String(string(b)),
+			PolicyName:     policyName,
+			RoleName:       aws.String(roleName),
+		}); err != nil {
+			return fmt.Errorf("failed to create inline policy for role %s: %w", role, err)
+		}
+
+		profileName := aws.String(fmt.Sprintf("%s-%s-profile", infraID, role))
+		if _, err := svc.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
+			var awsErr awserr.Error
+			if errors.As(err, &awsErr) && awsErr.Code() != iam.ErrCodeNoSuchEntityException {
+				return fmt.Errorf("failed to get %s instance profile: %w", role, err)
+			}
+			// If the profile does not exist, create it.
+			if _, err := svc.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
+				InstanceProfileName: profileName,
+				Tags:                tags,
+			}); err != nil {
+				return fmt.Errorf("failed to create %s instance profile: %w", role, err)
+			}
+			if err := svc.WaitUntilInstanceProfileExistsWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
+				return fmt.Errorf("failed to wait for %s instance profile to exist: %w", role, err)
+			}
+
+			// Finally, attach the role to the profile.
+			if _, err := svc.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
+				InstanceProfileName: profileName,
+				RoleName:            aws.String(roleName),
+			}); err != nil {
+				return fmt.Errorf("failed to add %s role to instance profile: %w", role, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// getOrCreateRole returns the name of the IAM role to be used,
+// creating it when not specified by the user in the install config.
+func getOrCreateIAMRole(ctx context.Context, nodeRole, infraID, assumePolicy string, ic installconfig.InstallConfig, tags []*iam.Tag, svc *iam.IAM) (string, error) {
+	roleName := aws.String(fmt.Sprintf("%s-%s-role", infraID, nodeRole))
+
+	var defaultRole string
+	if dmp := ic.Config.AWS.DefaultMachinePlatform; dmp != nil && len(dmp.IAMRole) > 0 {
+		defaultRole = dmp.IAMRole
+	}
+
+	masterRole := defaultRole
+	if cp := ic.Config.ControlPlane.Platform.AWS; cp != nil && len(cp.IAMRole) > 0 {
```

Comment:
Won't this fail if there is not `controlPlane` stanza defined in the installconfig? I think you need:
```suggestion
	if cp := ic.Config.ControlPlane; cp != nil && cp.Platform.AWS != nil && len(cp.Platform.AWS.IAMRole) > 0 {
```

Reply:
I wondered the same thing, but in my testing it didn't fail. So it seems like `ControlPlane` is always initialized.

That said, I still think it's good to check and will add it. :+1: